### PR TITLE
feat: provide endpoint to lock/unlock workspace

### LINKED ISSUE #8171: workspace actions: create endpoint to lock/unlock workspace
An endpoint is needed so that admins can unlock workspaces that may have been locked due to an inactivity timer set on a template. 

e.g. `PUT /api/v2/workspaces/<id>/lock`

### PR DESCRIPTION
**Pull request by sreya**
This PR adds an endpoint for locking/unlocking a workspace. For the sake of simplicity the current implementation allows a user to unlock their own workspace. This work is done in anticipation of #8172.
fixes #8171

### CODE REVIEW

File: coderd/rbac/object.go [hunk 1]
```diff
@@ -28,6 +28,18 @@ var (
...
 	}
 
+	ResourceWorkspaceBuild = Object{
+		Type: "workspace_build",
+	}
+
+	// ResourceWorkspaceLocked CRUD.
+	//	create/delete = make or delete workspaces
+	// 	read = access workspace
+	//	update = edit workspace variables
```
Comment (Emyrk) [inline coderd/rbac/object.go:38]: Delete these lines?

File: coderd/rbac/object.go [hunk 2]
```diff
@@ -28,6 +28,18 @@ var (
 		Type: "workspace",
 	}
 
+	ResourceWorkspaceBuild = Object{
+		Type: "workspace_build",
+	}
```
Comment (Emyrk) [inline coderd/rbac/object.go:33]: Comment what this is

File: coderd/workspacebuilds.go [hunk 3]
```diff
@@ -350,6 +350,11 @@ func (api *API) postWorkspaceBuilds(rw http.ResponseWriter, r *http.Request) {
 	)
 	var buildErr wsbuilder.BuildError
 	if xerrors.As(err, &buildErr) {
+		var authErr dbauthz.NotAuthorizedError
+		if xerrors.As(err, &authErr) {
+			buildErr.Status = http.StatusUnauthorized
+		}
+
```
Comment (Emyrk) [inline coderd/workspacebuilds.go:357]: :+1:

File: coderd/workspaces.go [hunk 4]
```diff
@@ -755,6 +751,61 @@ func (api *API) putWorkspaceTTL(rw http.ResponseWriter, r *http.Request) {
...
+			Message: "Internal error updating workspace locked status.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+
+	// TODO should we kick off a build to stop the workspace if it's started
+	// from this endpoint? I'm leaning no to keep things simple and kick
+	// the responsibility back to the client.
+	httpapi.Write(ctx, rw, code, resp)
```
Comment (Emyrk) [inline coderd/workspaces.go:806]: :+1: Punting sounds good to me

File: coderd/rbac/roles.go [hunk 5]
```diff
@@ -121,7 +121,7 @@ func ReloadBuiltinRoles(opts *RoleOptions) {
 		opts = &RoleOptions{}
 	}
 
-	var ownerAndAdminExceptions []Object
+	ownerAndAdminExceptions := []Object{ResourceWorkspaceLocked}
```
Comment (Emyrk) [inline coderd/rbac/roles.go:124]: We might want to keep `read` access here. Should users be able to read a locked workspace? So they can see it is locked?

**Review by Emyrk — COMMENTED** [review]
This is as discussed.

The RBAC enforcement is nice, I think we just need to add `read` access to locked workspaces where appropriate.

One open question is now that we have the idea of a `WorksapceBuildResource`, we might want to attach things like build logs to that??

File: coderd/rbac/roles.go (hunk 5 above)
Comment (sreya) [reply coderd/rbac/roles.go:124]: Yup absolutely, def a bug

**Comment by sreya** [comment]
> One open question is now that we have the idea of a WorksapceBuildResource, we might want to attach things like build logs to that??

Yeah and things like ProvisionerJobs, etc. I saw TODOs above some of the queries and figured I'd try to touch as little as possibe.

**Comment by Emyrk** [comment]
> > One open question is now that we have the idea of a WorksapceBuildResource, we might want to attach things like build logs to that??
> 
> Yeah and things like ProvisionerJobs, etc. I saw TODOs above some of the queries and figured I'd try to touch as little as possibe.

Fair for this PR

**Review by Emyrk — APPROVED** [review]
The locked workspace only affects builds, not the actual workspace. So you keep your read on the workspace :+1:

File: coderd/rbac/roles.go (hunk 5 above)
Comment (Emyrk) [reply coderd/rbac/roles.go:124]: I was wrong, this only affects WorkspaceBuild. Not Workspace. So this is correct.

**Pull request merged**